storage/src/journal/fixed.rs

```diff
@@ -0,0 +1,671 @@
+                        let mut buf = vec![0u8; N + size_of::<u32>()];
+                        let item = blob.read_at(&mut buf, offset).await.map_err(Error::Runtime);
+                        let next_offset = offset + (size_of::<u32>() + N) as u64;
+                        match item {
+                            Ok(_) => match Self::verify_integrity(&buf) {
+                                Ok(item) => Some((
+                                    Ok((
+                                        pruned_items
+                                            + items_per_blob * index as u64
+                                            + offset / (size_of::<u32>() + N) as u64,
+                                        item,
+                                    )),
+                                    (index, blob, next_offset),
+                                )),
+                                Err(err) => Some((Err(err), (index, blob, next_offset))),
+                            },
+                            Err(err) => Some((Err(err), (index, blob, blob_len))),
+                        }
+                    },
+                )
+            })
+            .buffer_unordered(concurrency)
+            .flatten())
+    }
+
+    fn most_recent_blob_number(&self) -> u64 {
+        self.blobs.len() as u64 + self.oldest_unpruned_blob as u64 - 1
+    }
+
+    /// Allow the journal to prune items older than `min_item_position`. The journal may still
+    /// retain some of these items, for example if they are part of the most recent blob.
+    pub async fn prune(&mut self, min_item_position: u64) -> Result<(), Error> {
+        let mut new_oldest_blob = min_item_position / self.cfg.items_per_blob;
+        if new_oldest_blob <= self.oldest_unpruned_blob as u64 {
+            // nothing to prune
+            return Ok(());
+        }
+        // Make sure we never prune the most recent blob
+        let most_recent_blob = self.most_recent_blob_number();
+        if new_oldest_blob >= most_recent_blob {
+            new_oldest_blob = most_recent_blob;
+        }
+
+        let mut blob_index = self.oldest_unpruned_blob as u64;
+        while blob_index < new_oldest_blob {
+            let blob = self.blobs.remove(0);
+            // Close the blob and remove it from storage
+            blob.close().await?;
+            self.runtime
+                .remove(&self.cfg.partition, Some(&blob_index.to_be_bytes()))
+                .await?;
+            debug!(blob = blob_index, "pruned blob");
+            self.pruned.inc();
+            self.tracked.dec();
+            blob_index += 1;
+        }
+        self.oldest_unpruned_blob = new_oldest_blob as usize;
+        Ok(())
+    }
+
+    /// Close the journal
+    pub async fn close(self) -> Result<(), Error> {
+        for (i, blob) in self.blobs.into_iter().enumerate() {
+            blob.close().await?;
+            debug!(blob = i + self.oldest_unpruned_blob, "closed blob");
```

Comment:
To avoid all these conversions/tracking, it might be easier to just store the blob number with the blob in a `BTreeMap`?

It would let you efficiently iterate in-order and know what blob you are interacting with.

Reply:
done -- worked out nicely because BTreeMap has getters for both first & last element.